### DISCUSSION TITLE
Nuxt JS statically built site reporting as "slow" by Google? / Nuxt JS performance tips

### DISCUSSION THREAD

**Discussion by sts-ryan-holton** [nuxt/nuxt#9057] [Questions (Nuxt 2)]
### Versions

- nuxt: v2.14.4
- node: v12.18.3

### Reproduction

See the attached Config file for Nuxt JS

**nuxt.config.js**
```javascript
require('dotenv').config();
import axios from 'axios'
import getRoutes from './utils/getRoutes'

export default {
  /*
  ** Nuxt rendering mode
  ** See https://nuxtjs.org/api/configuration-mode
  */
  mode: 'universal',
  /*
  ** Nuxt target
  ** See https://nuxtjs.org/api/configuration-target
  */
  target: 'static',
  /*
  ** Env variables
  */
  env: {
    BASE_URL: process.env.BASE_URL || "https://domain-monitor.io",
    API_URL: process.env.API_URL || "http://127.0.0.1:8000",
    ONESIGNAL_PUSH_APP_ID: process.env.ONESIGNAL_PUSH_APP_ID || "",
    ONESIGNAL_SAFARI_WEB_ID: process.env.ONESIGNAL_SAFARI_WEB_ID || "",
    GA_ID: process.env.GA_ID || ""
  },
  /*
  ** Headers of the page
  ** See https://nuxtjs.org/api/configuration-head
  */
  head: {
    title: 'Domain Monitor',
    meta: [
      { charset: 'utf-8' },
      { name: 'viewport', content: 'width=device-width, initial-scale=1' },
      { hid: 'og:title', name: 'og:title', content: 'Domain Monitor' },
      { hid: 'description', name: 'description', content: 'Keep track of your expiring domains today for FREE with our FREE domain monitoring product.' },
      { hid: 'og:description', name: 'og:description', content: 'Keep track of your expiring domains today for FREE with our FREE domain monitoring product.' }
    ],
    link: [
      { rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' },
      { rel: 'stylesheet', href: 'https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;700&display=swap' },
      { rel: 'stylesheet', href: 'https://fonts.googleapis.com/icon?family=Material+Icons' }
    ]
  },
  /*
  ** Global CSS
  */
  css: [
    '@/assets/scss/domain-monitor.scss'
  ],
  /*
  ** Bootstrap Vue
  */
  bootstrapVue: {
    bootstrapCSS: false,
    bootstrapVueCSS: false
  },
  /*
  ** Plugins to load before mounting the App
  ** https://nuxtjs.org/guide/plugins
  */
  plugins: [
    { mode: 'client', src: '@/plugins/bootstrap-vue-icons' },
    { mode: 'client', src: '@/plugins/vue-axios' },
    { mode: 'client', src: '@/plugins/vee-validate' },
    { mode: 'client', src: '@/plugins/vue-moment' },
    { mode: 'client', src: '@/plugins/content-images' },
    { mode: 'client', src: '@/plugins/content-videos' }
  ],
  /*
  ** Auto import components
  ** See https://nuxtjs.org/api/configuration-components
  */
  components: true,
  /*
  ** Nuxt.js dev-modules
  */
  buildModules: [
    ['@nuxtjs/google-analytics', {
      id: process.env.GA_ID
    }]
  ],
  /*
  ** Nuxt.js modules
  */
  modules: [
    'bootstrap-vue/nuxt',
    '@nuxtjs/axios',
    '@nuxtjs/auth',
    '@nuxtjs/onesignal',
    '@nuxtjs/pwa',
    '@nuxt/content',
    '@nuxtjs/sitemap',
    ['@nuxtjs/component-cache', { maxAge: 1000 * 60 * 5 }] // 5 minutes
  ],
  /*
  ** Auth config
  */
  auth: {
    redirect: {
      login: '/account/login',
      logout: '/account/login',
      callback: '/account/login',
      home: '/account/domains'
    },
    strategies: {
      local: {
        login: { url: 'auth/login', method: 'post', propertyName: 'token' },
        logout: { url: 'account/logout', method: 'post' },
        user: { url: 'auth/user', method: 'get', propertyName: 'user' }
      }
    }
  },
  /*
  ** One Signal
  */
  oneSignal: {
    init: {
      appId: process.env.ONESIGNAL_PUSH_APP_ID,
      safari_web_id: process.env.ONESIGNAL_SAFARI_WEB_ID,
      allowLocalhostAsSecureOrigin: true,
      welcomeNotification: {
          disable: true
      }
    }
  },
  /*
  ** Axios module configuration
  ** See https://axios.nuxtjs.org/options
  */
  axios: {
    baseURL: process.env.API_URL
  },
  /*
  ** Build configuration
  ** See https://nuxtjs.org/api/configuration-build/
  */
  build: {
    extractCSS: true,

    extend (config, ctx) {
      const vueLoader = config.module.rules.find((rule) => rule.loader === 'vue-loader')
      vueLoader.options.transformAssetUrls = {
        video: ['src', 'poster'],
        source: 'src',
        img: 'src',
        image: 'xlink:href',
        'b-img': 'src',
        'b-img-lazy': ['src', 'blank-src'],
        'b-card': 'img-src',
        'b-card-img': 'img-src',
        'b-card-img-lazy': ['src', 'blank-src'],
        'b-carousel-slide': 'img-src',
        'b-embed': 'src'
      }
    }
  },
  /*
  ** Sitemap configuration
  ** See https://www.npmjs.com/package/@nuxtjs/sitemap#setup
  */
  sitemap: {
    hostname: process.env.BASE_URL, // https://www.yoursite.com
    routes() {
      return getRoutes();
    },
  }
}

```
### What is Expected?

A score better than the one reported by Google Page Speed insights since my Nuxt JS site is statically built and should be nice and speedy considering I'm using Cloudflare for caching and am using the `@nuxtjs/component-cache` module.

### What is actually happening?

My site's performance is being reported as slow by the [page insights](https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fdomain-monitor.io%2F&tab=mobile) tool provided by Google.

I'm looking for tips from the team here to find out what I can do to improve the performance of my site, am I missing something in the config file? My site isn't even that large?

**Comment by stefanoschrs** [comment]
Same issue here, really disappointed by the production performance!

**Comment by sts-ryan-holton** [comment]
@stefanoschrs any tips on how to improve site performance? Don't seem to be getting any answers as of yet here?

**Comment by TheAlexLichter** [comment]
While Nuxt.js gives you pretty decent perf out of the box, it is your responsibility to ensure it "stay like that" while adding functionalities, dependencies, global components and so on. 😋 

Every globally registered Vue plugin (for example) or component library has a price -- more JS and CSS that will be loaded.

**Comment by sts-ryan-holton** [comment]
@manniL I think most people understand that point, is there anything specific in Nuxt for improving performance?

**Comment by stefanoschrs** [comment]
@manniL I don't think it's related adding plugins etc because I've tested also a plain installation with multiple pages and it stacks up. 

@sts-ryan-holton I haven't checked it properly yet but when opening the main page of [Skillfill](https://skillfill.ai) for example I can see that it's loading the styles and scripts of the other 4 pages (/evaluate, /enterprise, /data-facts, /about-us) as well which of course it slows down everything.

![image](https://user-images.githubusercontent.com/4215699/101597321-dac4ca00-39f6-11eb-8ec3-40d0a92fe1b9.png)

**Comment by sts-ryan-holton** [comment]
Any update here?

**Comment by larzon83** [comment]
@sts-ryan-holton @stefanoschrs are you talking about perceived perf or just Lighthouse?
I had bad scores in Lighthouse too (using Google Chrome). But in a new incognito window it was fine. In my case, it was the `axe` Chrome extension (for accessibility testing) that was the problem...

**Comment by sts-ryan-holton** [comment]
@larzon83 I'm talking about what Google is reporting to me through the PageSpeed Insights Tool, I appreciate Nuxt JS loads a lot on initial page load, but I would've thought the numbers would be higher.

**Comment by larzon83** [comment]
@sts-ryan-holton and you tried with incognito window without extensions?

**Comment by AdrianoCahete** [comment]
> I haven't checked it properly yet but when opening the main page of Skillfill for example I can see that it's loading the styles and scripts of the other 4 pages (/evaluate, /enterprise, /data-facts, /about-us) as well which of course it slows down everything.

I don't know how you are doing but for me, whats helped was:
 - Using scoped styles for components and one small "base" stylesheet with the most common classes. The base I call directly on the layout.
 - Add svgs as [inline sprite](https://css-tricks.com/svg-sprites-use-better-icon-fonts/) (even the icons)
 - Avoid calling external assets (fonts, icons, everything is local).
 - And run the build with the [`modern` flag](https://nuxtjs.org/docs/2.x/configuration-glossary/configuration-modern)


The worst page stats for my site is [83](https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fadrianocahete.dev%2Fui&tab=desktop). The best is [95](https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fadrianocahete.dev%2F&tab=desktop).

I know that my site isn't a big webpage, but it's comparable to a landing page. For bigger websites, I don't think that they can come closer to 90.

**Comment by StephanGerbeth** [comment]
Perhaps this module will help you to improve your performance issues.

https://github.com/GrabarzUndPartner/nuxt-speedkit

**Features**
- dynamic loading of viewport based page resources like fonts (subselectors, media queries), components, pictures
- optional loading prevention of resources at low bandwidth or weak hardware
- prevents the loading of unnecessary resources (including components) that are outside the current viewport.
- optional info layer concept to inform users about a reduced UX when bandwidth or hardware is compromised.

**Result**
- delivery of the minimum required resources based on the current viewport.
- lighthouse performance score of 100/100 (also on larger pages)

We are looking forward to the feedback.

**Comment by skorokithakis** [comment]
Just to chime in here, we have the same *perceived* performance issues. It seems to me like there's some sort of issue in how Google works with Nuxt, because Google thinks our site takes 10s to load, but it takes 1s on all browsers we've tested (obviously cleaned cache, loaded all resources, etc).

It seems to me like there isn't an *actual* issue, so there's nothing to "optimize", since the site loads basically instantly on any browser. The issue is that Google *thinks* that Nuxt is extremely slow. Here's the PageSpeed insights result:

![image](https://user-images.githubusercontent.com/23648/163976014-63f51c1c-705a-4518-9e35-aafaac960022.png)


This does not correspond to reality at all, as in Chrome the site achieves interactive in under two seconds. This really seems to be a Nuxt/Google issue.
